File: algorithms/Go/sorting/quicksort.go
/*
   Quick Sort is a divide and conquer algorithm.
   First we choose a pivot and split the array in two parts, one containing all elements less than or equal to the pivot and other contains the rest (the pivot element is in neither of them)
   Then we recursively sort the two arrays and finally concatenate them to get the sorted array.
   Average Time Complexity: O(n*log(n))
   For details explanation and proof of correctness check this -> https://cs.pomona.edu/classes/cs140/pdf/l09-quicksort-proof.pdf
*/

package sorting

import (
	"fmt"
	"math/rand"
	"time"
)

func quickSort(arr []int) []int {
	n := len(arr)
	if n <= 1 {
		return arr
	}
	rand.Seed(time.Now().UnixNano())
	pivot := rand.Int() % n
	var left []int
	var right []int
	for i := 0; i < n; i++ {
		if arr[i] <= arr[pivot] && i != pivot {
			left = append(left, arr[i])
		} else if arr[i] > arr[pivot] {
			right = append(right, arr[i])
		}
	}
	leftSorted := quickSort(left)
	rightSorted := quickSort(right)
	sortedArr := append(leftSorted, arr[pivot])
	sortedArr = append(sortedArr, rightSorted...)
	return sortedArr
}

func RunQuickSort() {
	arr := []int{10, 1, 6, 256, 2, 53, 235, 53, 1, 7, 23}
	fmt.Printf("Unsorted Array: %v\nSorted Array: %v", arr, quickSort(arr))
}

File: algorithms/C/sorting/quick-sort.c
// Quick Sort Algorithm in C

#include <stdio.h>

// Function to swap two elements
void swap(int *a, int *b)
{
    int temp = *a;
    *a = *b;
    *b = temp;
}

/* This function takes first element as pivot, places the pivot element at its correct position in sorted array, 
and places all smaller (smaller than pivot) to left of pivot and all greater elements to right of pivot.*/
int Partition(int a[], int low, int high)
{
    int pivot = a[low];
    int start = low;
    int end = high;

    while (start < end)
    {
        while (a[start] <= pivot)
            start++;
        while (a[end] > pivot)
            end--;
        if (start < end)
        {
            swap(&a[start], &a[end]);
        }
    }
    swap(&a[low], &a[end]);
    return end;
}

// Recursive function to sort an array using quick sort
void Quicksort(int a[], int low, int high)
{
    if (low < high)
    {
        int pivot = Partition(a, low, high);
        Quicksort(a, low, pivot - 1);
        Quicksort(a, pivot + 1, high);
    }
}

int main()
{
    int a[] = {5, 2, 4, 6, 1, 3};
    int n = sizeof(a) / sizeof(a[0]);
    Quicksort(a, 0, n - 1);
    for (int i = 0; i < n; i++){
        printf("%d ", a[i]);
    }
    return 0;
}

/*
Output:
1 2 3 4 5 6
*/